src/posit/connect/content.py

```diff
@@ -137,12 +140,231 @@ class ContentItem(Resource):
+        url = urls.append(self.config.url, f"v1/content/{self.guid}")
+        response = self.session.patch(url, json=body)
+        super().update(**response.json())
+
+    def _re_whatever(self) -> Task | None:
+        """Submit a re-whatever request (i.e., restart, refresh, etc).
+
+        A re-whatever is a catch-all term for restarting, refreshing, or re-whatever-ing the content requires to bounce it to a new state.
+
+        refresh:
+            For content that require variants. Find the default variant and render it again.
+
+        restart:
+            For content that require server threads. Toggle an unique environment variable and open the content, which activates a new server thread.
+
+        Returns
+        -------
+        Task | None:
+            A task for the content render when available, otherwise None
+
+        Raises
+        ------
+        RuntimeError
+            Found an incorrect number of default variants.
+
+        Examples
+        --------
+        >>> _re_whatever()
+        """
+        # Update the item to its current state.
+        # The 'app_mode' is not set until a bundle is created and deployed.
+        # During the deployment process, the 'app_mode' is read from manifest.json and written to the database.
+        # Until this occurs the 'app_mode' will be 'unknown'.
+        self.update()
+
+        if self.app_mode in {
+            "rmd-static",
+            "jupyter-static",
+            "quarto-static",
+        }:
+            variants = self._variants.find()
+            variants = [variant for variant in variants if variant.is_default]
+            if len(variants) != 1:
+                raise RuntimeError(
+                    f"Found {len(variants)} default variants. Expected 1. Without a single default variant, the content cannot be refreshed. This is indicative of a corrupted state."
+                )
+            variant = variants[0]
+            return variant.render()
+
+        if self.app_mode in {
+            "api",
+            "jupyter-voila",
+            "python-api",
+            "python-bokeh",
+            "python-dash",
+            "python-fastapi",
+            "python-shiny",
+            "python-streamlit",
+            "quarto-shiny",
+            "rmd-shiny",
+            "shiny",
+            "tensorflow-saved-model",
+        }:
+            random_hash = secrets.token_hex(32)
+            key = f"_CONNECT_RESTART_TMP_{random_hash}"
+            self.environment_variables.create(key, random_hash)
+            self.environment_variables.delete(key)
+            # GET via the base Connect URL to force create a new worker thread.
+            url = urls.append(dirname(self.config.url), f"content/{self.guid}")
+            self.session.get(url)
+            return None
+
+        warnings.warn(
+            f"Content '{self.guid}' with application mode '{self.app_mode}' does not require restarts"
+        )
+        return None
```

Comment:
@mmarchetti - I am adding you for this section in particular. This is the crux of the change, and I want to check if the business logic is reasonable.  The big caveat to this approach is upgrading the app_mode checks over time to accommodate new application modes. I've been tinkering with adding information to the warning message that directs users to open a GitHub issue as a lazy way to know if this happens.

Reply:
> The big caveat to this approach is upgrading the app_mode checks over time to accommodate new application modes.

We ran into [the same thing](https://github.com/rstudio/connectapi/pull/283#discussion_r1687197924) in connect_api. I wonder if we ought to have a more central place to retrieve this (either on the server, that enumerates what it knows about), or somewhere common in our code we could depend on. Of course, we don't need to / shouldn't solve that here, but it's not unique to this SDK!